src/order/height.lean

```diff
@@ -0,0 +1,276 @@
+  { obtain ⟨⟨l, hl⟩, hl'⟩ := enat.exists_le_of_supr_eq_top _ ha (n+1),
+    cases l,
+    { simp only [list.length, enat.coe_le_coe, nat.succ_ne_zero, nonpos_iff_eq_zero] at hl',
+      exact hl'.elim },
+    dsimp at *,
+    rw [enat.coe_le_coe, add_le_add_iff_right] at hl',
+    obtain ⟨l', h, h'⟩ := exists_chain_of_le_length hl hl',
+    rw ← h',
+    refine ⟨l_hd :: l', h, rfl⟩ },
+  casesI is_empty_or_nonempty {l : list α // chain' has_lt.lt l} with H H,
+  { rw [sup_chain_length, supr, set.range_eq_empty, Sup_empty, le_bot_iff] at hn,
+    exact (n.succ_ne_zero (enat.coe_inj.mp hn)).elim },
+  { obtain ⟨⟨l, hl⟩, h : ↑l.length = sup_chain_length α⟩ := enat.exists_eq_supr_of_ne_top _ ha,
+    cases l,
+    { simp only [← h, list.length, enat.coe_le_coe, nat.succ_ne_zero, nonpos_iff_eq_zero] at hn,
+      exact hn.elim },
+    have : n ≤ l_tl.length,
+    { rw [← h, enat.coe_le_coe] at hn, exact (add_le_add_iff_right 1).mp hn },
+    obtain ⟨l', h, h'⟩ := exists_chain_of_le_length hl this,
+    rw ← h',
+    refine ⟨l_hd :: l', h, rfl⟩ }
+end
+
+lemma enat.ne_zero_iff_one_le {n : enat} : n ≠ 0 ↔ 1 ≤ n :=
+begin
+  change ¬ n = ⊥ ↔ _,
+  rw [eq_bot_iff, ← lt_iff_not_le],
+  convert (enat.add_one_le_iff_lt (show (0 : enat) ≠ ⊤, from enat.coe_ne_top 0)).symm,
+  exact (zero_add 1).symm
+end
+
+lemma chain_height_eq_zero_iff (a : α) : chain_height a = 0 ↔ ∀ b, ¬ a < b :=
+begin
+  rw ← not_iff_not,
+  push_neg,
+  rw enat.ne_zero_iff_one_le,
+  split,
+  { intro h,
+    obtain ⟨l, hl, hl'⟩ := exists_chain_of_le_chain_height h,
+    obtain ⟨b, rfl⟩ := list.length_eq_one.mp hl',
+    cases hl,
+    exact ⟨b, by assumption⟩ },
+  { rintro ⟨b, r⟩,
+    refine le_trans _ (le_supr _ ⟨[b], by simpa⟩),
+    simp }
+end
+
+lemma sup_chain_length_eq_zero_iff  : sup_chain_length α = 0 ↔ is_empty α :=
+begin
+  rw [← not_iff_not, ← ne.def, enat.ne_zero_iff_one_le],
+  split,
+  { rintros h₁ h₂,
+    obtain ⟨⟨_|_, h⟩, h'⟩ := exists_chain_of_le_sup_chain_height h₁,
+    { simpa using h'.2 },
+    { exactI is_empty.false w_hd } },
+  { rw not_is_empty_iff,
+    rintro ⟨a⟩,
+    refine le_trans _ (le_supr _ ⟨[a], _⟩); simp }
+end
+
+lemma chain_height_le_of_lt {a b : α} (r : a < b) : chain_height b + 1 ≤ chain_height a :=
+begin
+  erw ← enat.supr_add,
+  apply supr_le,
+  rintro ⟨l, hl⟩,
+  refine le_trans _ (le_supr _ ⟨_, chain_cons.mpr ⟨r, hl⟩⟩),
+  simp
+end
+
+lemma chain_height_antitone : antitone (chain_height : α → enat) :=
+begin
+  intros a b r,
+  rcases le_iff_lt_or_eq.mp r with (h|rfl),
+  { exact (le_add_right $ le_of_eq rfl).trans (chain_height_le_of_lt h) },
+  { exact le_of_eq rfl }
+end
+
+lemma chain_height_le_sup_chain_length (a : α) :
+  chain_height a + 1 ≤ sup_chain_length α :=
+begin
+  erw ← enat.supr_add,
+  apply supr_le,
+  rintro ⟨l, hl⟩,
+  refine le_trans (le_of_eq _) (le_supr _ ⟨list.cons a l, hl⟩),
+  simp,
+end
+
+lemma sup_chain_length_eq_sup_chain_height :
+  sup_chain_length α = ⨆ a : α, chain_height a + 1 :=
```

Comment:
Note: we could probably simplify this lemma if we had some API for least strict upper bounds (the RHS is the least strict upper bound of the chain heights).